gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/handler/login/LoginCallbackParseHandler.java

```diff
@@ -132,14 +136,26 @@ private void restoreInitialQueryParams(RoutingContext context, Handler<AsyncResu
             handler.handle(Future.failedFuture(new InvalidRequestException("Missing state query param")));
             return;
         }
-
-        jwtService.decodeAndVerify(state, certificateManager.defaultCertificateProvider(), STATE)
+        var stateJwtCertProvider = certificateManager.defaultCertificateProvider();
+         jwtService.decodeAndVerify(state, stateJwtCertProvider, STATE)
                 .doOnSuccess(stateJwt -> {
                     final MultiMap initialQueryParams = RequestUtils.getQueryParams((String) stateJwt.getOrDefault(CLAIM_QUERY_PARAM, ""), false);
                     context.put(ConstantKeys.PARAM_CONTEXT_KEY, initialQueryParams);
                     context.put(ConstantKeys.PROVIDER_ID_PARAM_KEY, stateJwt.get(CLAIM_PROVIDER_ID));
                     context.put(ConstantKeys.REMEMBER_ME_PARAM_KEY, stateJwt.get(CLAIM_REMEMBER_ME));
 
+                    // TODO MRE PKCE-IDP make a constant, don't block
+
+                    stateJwtCertProvider.getProvider()
+                            .key()
+                            .map(k -> {
+                                var ecv = new String(Base64.getUrlDecoder().decode((String) stateJwt.get("ecv")));
+                                return CryptoUtils.decrypt(ecv, (Key) k.getValue());
+                            })
+                            .doOnSuccess(codeVerifier -> context.put("idp_code_verifier", codeVerifier))
+                            .ignoreElement()
+                            .blockingSubscribe();
```

Comment:
**issue**: please don't use blocking method in reactive chain. IMO we can manage the context update in a map/flapMap operator.